pmd-ui/src/main/java/net/sourceforge/pmd/util/fxdesigner/util/autocomplete/ResultSelectionStrategy.java

```diff
@@ -0,0 +1,249 @@
+
+        // Performs a left-to-right scan of the candidate string,
+        // trying to assign each of the chars of the query to a
+        // location in the string (also left-to-right)
+
+        // Score is computed a bit ad-hoc:
+        // +2 for a lonely char
+        // +10 for a character matching the start of a camelcase word (an uppercase char)
+        // the longer the submatch, the higher the match counts
+        // submatches occurring at the beginning of a word count more than in other places
+        // chars from the query that remain at the end penalise the score
+
+        // This algorithm is greedy and doesn't always select the best possible match result
+        // The second pass is even more unfair and allows to break ties
+
+        int candIdx = 0;  // current index in the candidate
+        int queryIdx = 0; // current index in the query
+        int score = 0;
+
+        // these are reset when a submatch ends
+        int lastMatchEnd = 0;
+        int curMatchStart = -1;
+        int matchLength = 0;
+        boolean isStartOfWord = true; // whether the current submatch is at the start of a camelcase word
+
+        TextFlow flow = new TextFlow(); // result
+
+        while (candIdx < candidate.length() && queryIdx < query.length()) {
+
+            char candChar = candidate.charAt(candIdx);
+            char queryChar = query.charAt(queryIdx);
+
+            if (Character.toLowerCase(candChar) == Character.toLowerCase(queryChar)) {
+                // it's the same char
+
+                matchLength++;
+
+                if (curMatchStart == -1) {
+                    // start of a match
+
+                    if (matchOnlyWordStarts && !isStartOfWord && !isWordStart(candidate, candIdx)) {
+                        // not the start of a word, don't record it as a match
+                        candIdx++;
+                        continue;
+                    }
+
+                    // set match start to current
+                    curMatchStart = candIdx;
+
+                    if (isWordStart(candidate, candIdx)) {
+                        // start of a match on the start of a word
+                        // e.g. query       coit
+                        //      candidate   ClassOrInterfaceType
+                        //                  ^    ^ ^ ^
+                        //      score       34
+
+                        isStartOfWord = true;
+                        score += 10;
+                    } else {
+                        isStartOfWord = false;
+                        score += 2;
+                    }
+
+                } else {
+                    // match is running-on
+
+                    // e.g. query       wur
+                    //      candidate   Würstchen
+                    //                  ^^^
+                    //      candidate   BratWurst
+                    //                      ^^^
+                    //      score       40 = 4 + 8 + 16 + (start of word : 10)
```

Comment:
4 + 8 + 16 + 10 = 38... Where do the extra 2 come from?

Reply:
That's embarrassing lol. You're obviously right, thanks :)